src/H5Dmpio.c

```diff
@@ -129,14 +118,21 @@ typedef struct H5D_chunk_alloc_info_t {
     H5F_block_t chunk_current;
     H5F_block_t chunk_new;
     hsize_t     chunk_idx;
+    haddr_t     dset_oloc_addr;
 } H5D_chunk_alloc_info_t;
 
 /*
  * Information for a chunk pertaining to the dataset's chunk
  * index entry for the chunk
  */
 typedef struct H5D_chunk_index_info_t {
-    hsize_t  chunk_idx;
+    /*
+     * These two fields must come in this order and next to
+     * each other for proper and efficient hashing
+     */
+    hsize_t chunk_idx;
+    haddr_t dset_oloc_addr;
```

Comment:
Since chunks being worked on could now have the same chunk index when they're from different datasets, the hash tables used by parallel compression make use of the chunk index and dataset object header address as the key. See https://troydhanson.github.io/uthash/userguide.html#_compound_keys for an example of using multiple struct fields as a compound key

Reply:
I've tried to figure this out on my own but it's taking too much time: are these structs always calloc'ed? Otherwise I'm concerned if there's a gap in the struct it could have garbage that would affect the key value.

Reply:
The structs are memset'ed during I/O setup at https://github.com/jhendersonHDF/hdf5/blob/parallel_comp_multi_dset/src/H5Dmpio.c#L3206-L3207. Then, the multi-field key approach is only used in one place in the new code for hashing and the code uses the structs directly (without them being copied into another temporary structure with possible garbage for example), so there shouldn't be any garbage in the structs during hashing. The same struct is nested inside the `H5D_chunk_insert_info_t` structure that is used for chunk index reinsertion (since we need the chunk index and other fields), so there could be garbage in that case, but that code uses a different hash table with a key based on a single field instead of the compound key approach, so we should be good.

Reply:
OK sounds good, we might want to add comments about this though, since this would be very hard to debug if someone changes it so those conditions no longer hold true